---
title: "R Notebook"
output: html_notebook
---
# htseq-count input
```{r}
directory <- "~/Documents/R/file_all"
```

```{r}
sampleFiles <- grep("lobe",list.files(directory),value=TRUE)
```

```{r}
sampleCondition <- sub("(.*lobe).*","\\1",sampleFiles)
```

```{r}
sampleTable <- data.frame(sampleName = sampleFiles,
                          fileName = sampleFiles,
                          condition = sampleCondition)
sampleTable$condition <- factor(sampleTable$condition)
```
# build the DESeqDataSet
```{r}
library("DESeq2")
dds <- DESeqDataSetFromHTSeqCount(sampleTable = sampleTable,
                                       directory = directory,
                                       design= ~ condition)
dds
```
# Pre-filtering
```{r}
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
```
# Note on factor levels
```{r}
dds$condition <- factor(dds$condition, levels = c("upperlobe","lowerlobe"))
```

```{r}
dds$condition <- droplevels(dds$condition)
```
# results table
```{r}
dds <- DESeq(dds)
```
```{r}
dds
```

# Independent hypothesis weighting (p value filtering)
```{r}
library("IHW")
res <- results(dds, filterFun=ihw, contrast=c("condition","upperlobe","lowerlobe"), alpha=0.05)
res
```
# remove the version number of the gene Ensembl number
```{r}
enemble_id <- substr(row.names(dds),1,15)
rownames(res) <- enemble_id
```
# change Ensembl number of gene in dataframe to gene id
## add a column 
```{r}
RawCounts <- res
Ensembl_ID <- data.frame(Ensembl_ID = row.names(RawCounts))
rownames(Ensembl_ID) <- Ensembl_ID[,1]
RawCounts <- cbind(Ensembl_ID,RawCounts)
```
## create a file to associate the ensembl id and gene id
```{r}
get_map = function(input) {
  if (is.character(input)) {
    if(!file.exists(input)) stop("Bad input file.")
    message("Treat input as file")
    input = data.table::fread(input, header = FALSE)
  } else {
    data.table::setDT(input)
  }
  
  input = input[input[[3]] == "gene", ]
  
  pattern_id = ".*gene_id \"([^;]+)\";.*"
  pattern_name = ".*gene_name \"([^;]+)\";.*"
  
  gene_id = sub(pattern_id, "\\1",input[[9]])
  gene_name = sub(pattern_name, "\\1", input[[9]])
  
  Ensembl_ID_TO_Genename <- data.frame(gene_id = gene_id, gene_name = gene_name, stringsAsFactors = FALSE)
  return(Ensembl_ID_TO_Genename)
}
Ensembl_ID_TO_Genename <- get_map("gencode.v38.basic.annotation.gtf")

gtf_Ensembl_ID <- substr(Ensembl_ID_TO_Genename[,1],1,15)
Ensembl_ID_TO_Genename <- data.frame(gtf_Ensembl_ID, Ensembl_ID_TO_Genename[,2])
colnames(Ensembl_ID_TO_Genename) <- c("Ensembl_ID","gene_id")
write.csv(Ensembl_ID_TO_Genename,file = "Ensembl_ID_TO_Genename.csv")
```
#replace
```{r}
res_g <-merge(Ensembl_ID_TO_Genename,RawCounts,by="Ensembl_ID")
```
## remove unnecessary columns and duplicate gene ids
```{r}
res_g <- res_g[order(res_g[,"gene_id"]),]
index <- duplicated(res_g$gene_id)
res_g <- res_g[!index,]
rownames(res_g) <- res_g[,"gene_id"]
res_g <- res_g[,-c(1:2)]
```
# check
```{r}
head(res_g)
```
# effect size shrinkage
```{r}
resultsNames(dds)
```
## takes long time
```{r}
resLFC <- lfcShrink(dds,coef="condition_lowerlobe_vs_upperlobe", type="apeglm")
resLFC
```
# speed-up and parallelization thoughts
```{r}
library("BiocParallel")
register(MulticoreParam(4))
```
# order results table by p-value
```{r}
resOrdered <- res_g[order(res_g$pvalue),]
```

```{r}
sum(res_g$padj < 0.05, na.rm=TRUE)
```

# Exporting results
## MA-plot
```{r}
plotMA(res, ylim=c(-2,2))
```
# more useful
```{r}
plotMA(resLFC, ylim=c(-2,2))
```

```{r}
resultsNames(dds)
```
# 3 shrinkage estimator
## resAsh takes long time
```{r}
resNorm <- lfcShrink(dds, coef=2, type="normal")
resAsh <- lfcShrink(dds, coef=2, type="ashr")
```

```{r}
par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
plotMA(resLFC, xlim=xlim, ylim=ylim, main="apeglm")
plotMA(resNorm, xlim=xlim, ylim=ylim, main="normal")
plotMA(resAsh, xlim=xlim, ylim=ylim, main="ashr")
```
# plot counts
```{r}
plotCounts(dds, gene=which.min(res$padj), intgroup="condition")
```

```{r}
d <- plotCounts(dds, gene=which.min(res$padj), intgroup="condition", 
                returnData=TRUE)
library("ggplot2")
ggplot(d, aes(x=condition, y=count)) + 
  geom_point(position=position_jitter(w=0.1,h=0)) + 
  scale_y_log10(breaks=c(25,100,400))
```
# more information
```{r}
mcols(res)$description
```

# finally write csv
```{r}
getwd()
write.csv(as.data.frame(res_g), 
          file="res_g.csv")
```
# differential expression gene p<0.05
```{r}
resSig_0.05 <- subset(resOrdered, padj < 0.05)
resSig_0.05[which(resSig_0.05$log2FoldChange > 0), "up_down"] <- "up"
resSig_0.05[which(resSig_0.05$log2FoldChange < 0), "up_down"] <- "down"
resSig_0.05
```

```{r}
getwd()
write.csv(as.data.frame(resSig_0.05), 
          file="differential_expression.csv")
```

# data transformations and visualization
## vsd data transformations
```{r}
vsd <- vst(dds, blind=FALSE)
```
## rld data transformations
## takes too much time and have no results
```{r}
rld <- rlog(dds, blind=FALSE)
```

```{r}
head(assay(vsd), 3)
```
## change of standard deviation
```{r}
# this gives log2(n + 1)
ntd <- normTransform(dds)
library("vsn")
meanSdPlot(assay(ntd))
```

```{r}
meanSdPlot(assay(vsd))
```
### need rlg
```{r}
meanSdPlot(assay(rld))
```
# data quality assessment by sample clustering and visualization
## heatmap of ntd
### show_colnames = FALSE
```{r}
library("pheatmap")
select <- order(rowMeans(counts(dds,normalized=TRUE)),
                decreasing=TRUE)[1:20]
df <- as.data.frame(colData(dds)[,c("condition", "sizeFactor")])
pheatmap(assay(ntd)[select,], cluster_rows=FALSE, show_rownames=FALSE,
         cluster_cols=FALSE, annotation_col=df, show_colnames = FALSE)
```
## heatmap of vsd
```{r}
pheatmap(assay(vsd)[select,], cluster_rows=FALSE, show_rownames=FALSE,
         cluster_cols=FALSE, annotation_col=df, show_colnames = FALSE)
```
## heatmap of rlg
### need rlg
```{r}
pheatmap(assay(rld)[select,], cluster_rows=FALSE, show_rownames=FALSE,
         cluster_cols=FALSE, annotation_col=df)
```
## Heatmap of the sample-to-sample distances
```{r}
sampleDists <- dist(t(assay(vsd)))
```
### show_rownames = FALSE
```{r}
library("RColorBrewer")
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$condition, vsd$type, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors, show_rownames = FALSE)
```
## principal component plot of the samples
```{r}
plotPCA(vsd, intgroup="condition")
```
## same thing
```{r}
pcaData <- plotPCA(vsd, intgroup="condition", returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color=condition)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed()
```

